cfgov/research_reports/models.py

```diff
@@ -0,0 +1,235 @@
+import os
+
+from django.conf import settings
+from django.core import management
+from django.db import models
+from django.utils.safestring import mark_safe
+
+from wagtail.admin.edit_handlers import (
+    FieldPanel, InlinePanel, MultiFieldPanel, ObjectList, TabbedInterface
+)
+from wagtail.admin.forms.models import WagtailAdminModelFormMetaclass
+from wagtail.admin.forms.pages import WagtailAdminPageForm
+from wagtail.core.models import PageManager
+from wagtail.documents.edit_handlers import DocumentChooserPanel
+
+from modelcluster.fields import ParentalKey
+from modelcluster.models import ClusterableModel
+
+from v1.models import SublandingFilterablePage
+from v1.models.base import CFGOVPage
+
+
+alpha_map = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+
+
+def get_report_parts(is_appendix=False):
+    def format(s, *args):
+        if(is_appendix):
+            return s.format(*[alpha_map[arg] for arg in args])
+        else:
+            return s.format(*[arg + 1 for arg in args])
+
+    def sec(request, page):
+        sections = [section for section in
+                    page.report_sections.all().order_by('pk')
+                    if section.is_appendix == is_appendix]
+
+        return [{
+            'expanded': False,
+            'title': section.header,
+            'body': section.body,
+            'url': format('#section-{}', i),
+            'numbering': format('{}: ' if is_appendix else '{}. ', i),
+            'children': [{
+                'title': subsection.header,
+                'body': subsection.body,
+                'url': format('#section-{}.{}', i, j),
+                'numbering': '' if is_appendix else format('{}.{} ', i, j),
+                'children': [{
+                    'title': section3.header,
+                    'body': section3.body,
+                    'url': format('#section-{}.{}.{}', i, j, k),
+                    'numbering': '' if is_appendix else
+                                 format('{}.{}.{} ', i, j, k),
+                } for k, section3 in enumerate(
+                    subsection.report_sections_level_three.all().order_by('pk')
+                )]
+            } for j, subsection in enumerate(
+                section.report_subsections.all().order_by('pk')
+            )]
+        } for i, section in enumerate(sections)]
+
+    return sec
+
+
+get_report_sections = get_report_parts()
+get_report_appendices = get_report_parts(True)
+
+
+def get_researchers():
+    return dict([
+        (r.title, r.url) for r in
+        SublandingFilterablePage.objects.get(pk=4833).get_children()
+    ])
+
+
+def _get_deploy_environment():
+    return getattr(settings, 'DEPLOY_ENVIRONMENT', 'local')
+
+
+class ReportSection(ClusterableModel):
+    header = models.CharField(max_length=200, blank=True)
+    is_appendix = models.BooleanField(default=False)
+    body = models.TextField(blank=True)
+    panels = [
```

Comment:
Minor style thing, but Wagtail convention [seems to be](https://docs.wagtail.io/en/v2.9.3/topics/pages.html#an-example-wagtail-page-model) to put panel definitions below all of the model fields. I find that helpful to separate out database fields from other attributes on the class.

Reply:
I see. I had grouped things by which tab they appeared on in the Wagtail admin, since that organization was helpful for me, but I can move them to match the standard.

Reply:
Fixed in 109a63e.